worker/src/main/java/gov/cms/ab2d/worker/adapter/bluebutton/BfdClientAdapterImpl.java

```diff
@@ -0,0 +1,51 @@
+package gov.cms.ab2d.worker.adapter.bluebutton;
+
+import gov.cms.ab2d.bfd.client.BFDClient;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.hl7.fhir.dstu3.model.Bundle;
+import org.hl7.fhir.dstu3.model.Bundle.BundleEntryComponent;
+import org.hl7.fhir.dstu3.model.Resource;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.scheduling.annotation.AsyncResult;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class BfdClientAdapterImpl implements BfdClientAdapter {
+
+    @Autowired
+    private BFDClient bfdClient;
+
+    @Async("bfd-client")
+    @Override
+    public Future<List<Resource>> getEobBundleResources(String patientId) {
```

Comment:
So this is the _async_ method that deals with a single patient's EOBs. I _think_, rather than trying to sort out the Futures in the "main" job Thread, you could just do the the entire per-patient flow right here:

- Get the EOBs from BFD
- Serialize them into JSON
- Append to the `ndjson` file. Need to pay attention to the fact that we have _one_ file and _many_ threads writing to it.

So the "main" job thread would just have to wait until all Futures are done and move the job to `COMPLETED`. You should end up with a cleaner code using this approach.

Reply:
```Need to pay attention to the fact that we have one file and many threads writing to it.```
The way I solved the contention during writing to the file is by doing the writing to the ndjson file into the main thread which being single threaded would not have the contention issue.

Reply:
Right. `java.util.concurrent.locks.ReentrantLock` should help do the same with the new approach.

Reply:
I've attempted to refactor as described above where the output and error files would be written to from inside the async method.